app/controllers/collections_controller.rb

```diff
@@ -121,6 +121,60 @@ def revoke_access
     redirect_to licences_path
   end
 
+  def add_items_to_collection
+    # referenced documents (HCSVLAB-1019) are already handled by the look_for_documents part of the item ingest
+    begin
+      # Perform parameter transformations and request validations
+      params[:items] = JSON.parse(params[:items]) if params[:items].is_a? String
+      params[:file] = [params[:file]] unless params[:file].nil? or params[:file].is_a? Array
+      validate_add_items_request(params)
+
+      collection = Collection.find_by_name(params[:id])
+      corpus_dir = corpus_dir(params[:id])
+      items = []
+      items_added = []
+      uploaded_files = []
+
+      # Upload each of the files provided in the multipart HTTP request
+      if !params[:file].nil?
+        params[:file].each do |uploaded_file|
+          uploaded_files.push(upload_document_using_multipart(File.join(corpus_dir, uploaded_file.original_filename), uploaded_file, collection.name))
+        end
+      end
+
+      # Update document source in graph, convert metadata to RDF and ingest each item
```

Comment:
This method is still doing a lot of work, you could split the rest into a few extra methods, 
```
def add_items_to_collection
  begin
    cleans_params(params)
    validate_add_items_request(params)
    process_uploaded_files
    process_items # which probably has a process_documents helper
    ingest_items
resuce
    ...
```
If you have good names for variables and methods, the advantage of having lots of little methods over one big one is that it reads like pseudocode, so it is very easy for someone else to understand. Even if it was completely broken, someone else can come along and understand what the intent was. It also makes it easier to track down bugs because they can be isolated to smaller sections of code.